test: client compat migrated to backend integration tests and fixes

merlin-northern opened the pull request:
test: client compat migrated to backend integration tests and test separation

Client compatibility test is re-implemented as a backend integration test.

A general mechanism to run tests in a separate and completely independent
session is introduced; all that is needed is to name the test file with
`+separate` before .py extension, for instance moving `test_artifact.py`
to a separate pytest session can be achieved by changing the file name
to `test_artifact+separate.py`.

Ticket: QA-683
Signed-off-by: Peter Grzybowski <peter@northern.tech>

merlin-northern commented:
@kjaskiewiczz let's have a romantic chat about it. there are sone parts that are better spoken about. :)

alfrunes commented on backend/tests/integration/run:
```diff
@@ -1,5 +1,50 @@
+        esac
+    fi
+    return $rc
+}
```
Can we simply replace `docker-compose` with `docker compose`?
This is adding unnecessary amount of complexity and the docker compositions will not work with out compose v2 (plugin) anyway.

merlin-northern replied:
sure. I added a variable in case there is another docker compose in the future (or past :))

alfrunes commented on backend/tests/Dockerfile.integration:
```diff
@@ -2,6 +2,7 @@ FROM python:3.13-bullseye
 
 ARG K8S_VERSION=1.29
 
+RUN sed -i.backup -e 's/^deb /deb [trusted=yes] /' /etc/apt/sources.list
```
What is this for? Does this need a separate commit?

merlin-northern replied:
without it the tests dont run for me on any env including a fresh new debian machine with nothing but docker installed, since it needs signature verified and in order to have it needs to install something from a repo that is not verified, classical. moved to separate commit.

alfrunes commented on backend/tests/integration/README.md:
Great that you're adding documentation!
Could you please move this into a separate commit?

merlin-northern replied:
sure. moved.

alfrunes commented on backend/tests/integration/tests/test_compat+separate.py:
```diff
@@ -0,0 +1,613 @@
+    """
+    mongo_cleanup(mongo)
+    yield mongo
+    mongo_cleanup(mongo)
```
Isn't this just duplicating `clean_mongo`?

merlin-northern replied:
general comment for all the "unused" comments from you: I am sorry it was a left over in the file, I cleaned up everything. thank you for keeping an eye.

alfrunes commented on backend/tests/integration/tests/test_compat+separate.py:
Most fixtures defined here seem to duplicate entire or partial fixtures defined elsewhere. Can we try to collect at least the common ones in the local `conftest.py` to avoid all this boilerplate?

alfrunes replied:
On second passing, only the `user` fixture is used by `test_compat`, which means that of the non-test code, only `clean_mongo_client` (`clean_mongo`), `clean_migrated_mongo` and `user` is used. The rest seems to be unused.

merlin-northern replied:
thank you see above.

alfrunes commented on backend/tests/integration/tests/test_compat+separate.py:
```diff
@@ -0,0 +1,613 @@
+            assert r.status_code == 204
+
+        m = MongoClient("mender-mongo:27017")
+        m.cleanup()
```
This is handled by the fixture

merlin-northern replied:
thank you see above.

alfrunes commented on backend/tests/integration/tests/test_compat+separate.py:
```diff
@@ -0,0 +1,613 @@
+
+
+@pytest.fixture(scope="function")
+def tenants_with_plans(clean_mongo):
```
This seems unused

merlin-northern replied:
thank you see above.

alfrunes commented on backend/tests/integration/tests/test_compat+separate.py:
```diff
@@ -0,0 +1,613 @@
+    return devs[lo:hi]
+
+
+def compare_aset(authset, api_authset):
```
unused?

merlin-northern replied:
thank you see above.

alfrunes commented on backend/tests/integration/tests/test_compat+separate.py:
```diff
@@ -0,0 +1,613 @@
+    return dev
+
+
+def filter_and_page_devs(devs, page=None, per_page=None, status=None):
```
unused?

merlin-northern replied:
thank you see above.

alfrunes commented on backend/tests/integration/README.md:
````diff
@@ -0,0 +1,41 @@
+```
+
+where the `test_compat` corresponds to thetest file name without `.py` and `+separate`.
+
````
This is a lot of instructions for adding a test, and it is also adding complexity to the entire structure.
How about doing something like [this](https://docs.pytest.org/en/7.1.x/example/markers.html#custom-marker-and-command-line-option-to-control-test-runs)?
Or, add a pytest argument that select only that test and skip all others and use that in the docker compose override?

merlin-northern replied:
I will do anything to stay away from the pytest mechanisms as they are unportable and really, really obscure and make things _extremely_ confusing. so unless you block it, I will fight for this. all you do is you add "+separate" to the filename, how can this be complicated?

merlin-northern replied:
going "the pytest way" defies the purpose here: with the present patch we have a general mechanism to run any test separately.  all you need to do is add as many test_*+separate.py files and each one of them and out of the box will run completely isolated.

merlin-northern committed:
test: backend integration tests package installation fix.

Ticket: QA-683
Signed-off-by: Peter Grzybowski <peter@northern.tech>

merlin-northern committed:
docs: docmentation on backend integration tests running separately

Ticket: QA-683
Signed-off-by: Peter Grzybowski <peter@northern.tech>

merlin-northern committed:
test: client compat migrated to backend integration tests and test separation

Client compatibility test is re-implemented as a backend integration test.

A general mechanism to run tests in a separate and completely independent
session is introduced; all that is needed is to name the test file with
`+separate` before .py extension, for instance moving `test_artifact.py`
to a separate pytest session can be achieved by changing the file name
to `test_artifact+separate.py`.

Ticket: QA-683
Signed-off-by: Peter Grzybowski <peter@northern.tech>

alfrunes commented on backend/tests/integration/run:
```diff
@@ -272,6 +277,28 @@ for suite in "${TEST_SUITES[@]}"; do
+            compose_cmd logs >> "$tmppath" 2>&1
+        fi
+    done < <(find "${MENDER_SERVER_PATH}/backend/tests/integration/tests" -name '*+separate.py')
+
```
This makes it impossible to parallelize backend integration tests into separate pytest jobs.
Can we:
 * put this in a separate suite (`-s` argument)?
 * put this test in a separate test module and extend the docker-compose.backend-tests-compat.yml to mount this test module under `/tests`?

merlin-northern replied:
I do not understand. this makes it possible to run any test you put in the test_name+separate.py in a separate session. why do we have to complicate it?

alfrunes commented on backend/tests/integration/tests/test_compat+separate.py:
```diff
@@ -0,0 +1,200 @@
+        assert len(accepted) == len(expected_client_versions)
+
+        invm = ApiClient(inventory.URL_MGMT)
+        seen_versions = {}
```
```suggestion
        seen_versions = set()
```

alfrunes commented on backend/tests/integration/tests/test_compat+separate.py:
```diff
@@ -0,0 +1,200 @@
+                j = j - 1
+                time.sleep(1)
+            mender_client_version = mender_client_version_attributes[0]["value"]
+            seen_versions[mender_client_version] = True
```
```suggestion
            seen_versions.add(mender_client_version)
```

alfrunes commented on backend/tests/integration/tests/test_compat+separate.py:
```diff
@@ -0,0 +1,200 @@
+
+        artifact_file = "tests/data/date.mender"
+        for i in range(len(expected_client_versions)):
+            assert seen_versions[expected_client_versions[i]]
```
This assertion will never fail because it throws a `KeyError` instead.
```suggestion
            assert expected_client_versions[i] in seen_versions
```

alfrunes commented on backend/tests/integration/tests/test_compat+separate.py:
```diff
@@ -0,0 +1,200 @@
+            "2.5.4",
+            "2.2.1",
+            "2.1.3",
+            "3.3.2",
```
There are versions here that are not present in the composition: for instance: 4.0.6

merlin-northern replied:
yeah this is tricky, this one is hidded in the docker client tag 3.7

alfrunes committed:
refac: Move compat tests to dedicated test suite

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes commented:
I did this to be able to run the test it in isolation: https://github.com/merlin-northern/mender-server/compare/compat_test_in_backend_integration...alfrunes:mender-server:compat_test_in_backend_integration

merlin-northern commented:
> I did this to be able to run the test it in isolation: [merlin-northern/mender-server@compat_test_in_backend_integration...alfrunes:mender-server:compat_test_in_backend_integration](https://github.com/merlin-northern/mender-server/compare/compat_test_in_backend_integration...alfrunes:mender-server:compat_test_in_backend_integration)

I see. and if I have a new test I want to run in separation, what do I do?

merlin-northern committed:
Merge pull request #1 from alfrunes/compat_test_in_backend_integration

refac: Move compat tests to dedicated test suite

alfrunes committed:
test(integration): Use built-in PYTEST_ADDOPTS in run script

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

alfrunes commented:
Sorry, after cleaning up my PR to your fork, I accidentally broke the other test suites. https://github.com/merlin-northern/mender-server/pull/2

alfrunes committed:
test(integration): Restore conftest.py to test modules

Signed-off-by: Alf-Rune Siqveland <alf.rune@northern.tech>

merlin-northern committed:
Merge pull request #2 from alfrunes/compat_test_in_backend_integration

test(integration): Use built-in PYTEST_ADDOPTS in run script

merlin-northern commented:
> Sorry, after cleaning up my PR to your fork, I accidentally broke the other test suites. [merlin-northern#2](https://github.com/merlin-northern/mender-server/pull/2)

do not worry

merlin-northern commented:
@alfrunes https://github.com/mendersoftware/mender-server-enterprise/pull/255 is the enterprise counter part.
